source/FreeRTOS_IGMP.c

```diff
@@ -0,0 +1,1068 @@
+                    vPortFree( pxMRD );
+                    pxMRD = NULL;
+                }
+            }
+        #endif /* ( ipconfigUSE_IPv6 == 1) */
+    #endif /* ipconfigUSE_LLMNR */
+
+    #if ( ipconfigUSE_MDNS == 1 )
+        if( NULL != ( pxMRD = ( MCastReportData_t * ) pvPortMalloc( sizeof( MCastReportData_t ) ) ) )
+        {
+            listSET_LIST_ITEM_OWNER( &( pxMRD->xListItem ), ( void * ) pxMRD );
+            /* Quick and dirty assignment of end-point. This will probably have to be re-designed and re-done. */
+            pxMRD->pxEndPoint = NULL;
+            pxMRD->xMCastGroupAddress.xIs_IPv6 = pdFALSE_UNSIGNED;
+            pxMRD->xMCastGroupAddress.xIPAddress.ulIP_IPv4 = ipMDNS_IP_ADDRESS;
+            BaseType_t bReportItemConsumed = xAddIGMPReportToList( pxMRD );
+
+            if( pdTRUE != bReportItemConsumed )
+            {
+                /* This should not happen, but if it does, free the memory that was used */
+                vPortFree( pxMRD );
+                pxMRD = NULL;
+            }
+        }
+
+        #if ( ipconfigUSE_IPv6 == 1 )
+            if( NULL != ( pxMRD = ( MCastReportData_t * ) pvPortMalloc( sizeof( MCastReportData_t ) ) ) )
+            {
+                listSET_LIST_ITEM_OWNER( &( pxMRD->xListItem ), ( void * ) pxMRD );
+                /* Quick and dirty assignment of end-point. This will probably have to be re-designed and re-done. */
+                pxMRD->pxEndPoint = NULL;
+                pxMRD->xMCastGroupAddress.xIs_IPv6 = pdTRUE_UNSIGNED;
+                memcpy( ( void * ) &pxMRD->xMCastGroupAddress.xIPAddress.xIP_IPv6.ucBytes[ 0 ], &ipMDNS_IP_ADDR_IPv6.ucBytes[ 0 ], sizeof( IPv6_Address_t ) );
+                BaseType_t bReportItemConsumed = xAddIGMPReportToList( pxMRD );
+
+                if( pdTRUE != bReportItemConsumed )
+                {
+                    /* This should not happen, but if it does, free the memory that was used */
+                    vPortFree( pxMRD );
+                    pxMRD = NULL;
+                }
+            }
+        #endif /* ( ipconfigUSE_IPv6 == 1) ) */
+    #endif /* ( ipconfigUSE_MDNS == 1) */
+
+    vIGMPTimerReload( pdMS_TO_TICKS( ipIGMP_TIMER_PERIOD_MS ) );
+}
+
+/**
+ * @brief Process an IGMP packet.
+ *
+ * @param[in,out] pxIGMPPacket: The IP packet that contains the IGMP message.
+ *
+ * @return eReleaseBuffer This function usually returns eReleaseBuffer as IGMP reports are
+ * scheduled for later. If however the user has implemented IGMP Snooping, the return is
+ * controlled by the eApplicationIgmpFrameReceivedHook function. That function might return
+ * eFrameConsumed if it decided to forward the frame somewhere.
+ */
+eFrameProcessingResult_t eProcessIGMPPacket( NetworkBufferDescriptor_t * const pxNetworkBuffer )
+{
+    eFrameProcessingResult_t eReturn = eReleaseBuffer;
+
+    if( pxNetworkBuffer->xDataLength < sizeof( IGMPPacket_t ) )
+    {
+        return eReturn;
```

Comment:
Can we have single return statement per function?

Reply:
@tony-josi-aws Yes, I'm aware of that and it is on my to-do list to go over things like returns, naming, casting, etc

Reply:
I think we can help out on the MISRA violations once the code is stable.